docs/contract/confidential-smart-contract.md

```diff
@@ -0,0 +1,264 @@
+# Confidential Smart Contract
+
+In this chapter we are going to see how to:
+1. write a smart contract which stores and loads data to and from a
+   confidential store and
+2. instantiate and call the smart contract without revealing the call arguments.
+
+## PublicCell VS ConfidentialCell
+
+In the [hello world](./hello-world.md) example we used
+`oasis_contract_sdk_storage::cell::PublicCell<T>` to access the key-value store
+of that contract instance. In this case the value was stored unencrypted on the
+blockchain associated with the hash of the key we provided in the constructor.
+
+Cipher supports another primitive
+`oasis_contract_sdk_storage::cell::ConfidentialCell<T>` which allows you to
+access the data of the contract instance confidentially assured by a
+hardware-level encryption.
+
+In contrast to non-confidential calls the hash of the key we passed to the
+constructor, the runtime library also computes a nonce each time the data is
+accessed. This nonce ensures the confidentiality so that the accessed key
+appears different each time from the blockchain observer and even from the node
+operator. It is generated from:
+- the round number,
+- the number of the call inside the current smart contract execution,
+- the number of the smart contract executions until the contract was
+  instantiated.
+
+As an additional safety measure, you will also need to provide the
+instance of the *confidential store* of the contract instance. This is obtained
+by calling `confidential_store()` of the contract's *context* object. If, for
+example, the node operator tries to execute your code in a non-confidential
+environment, this method will panic.
```

Comment:
It would be better to say that in a non-confidential environment the node operator will not be able to obtain the keys required to perform decryption so the operation would fail.